### DISCUSSION TITLE
Build Docker image for patched/forked renovate

### DISCUSSION THREAD

**Discussion by skral** [renovatebot/renovate#10918] [Request Help]
Hi,

I am struggling to find the best way to build a Docker image for a patched renovate. I read through discussions and the Dockerfiles of a view repositories but couldn't find an obvious answer, unfortunately.

Context:
I am using a Python package manager called [rez](https://github.com/nerdvegas/rez) which isn't supported out of the box but I configured renovate with some regular expressions to be able to recognize dependencies. I did have to add "rez" versioning though and so I went into `lib/versioning` and created a `rez` folder. Got it all working on my dev machine with Visual Studio Code and the dev container.

I now have:
- A `rez` folder in `lib/versioning/
- Two lines of changes in `lib/versioning/api.ts` to add the `rez` versioning api

What I would like to achieve:
Have a Docker image of the patched renovate understanding `rez` versioning that I can run self hosted on my GitLab instance.

My current approach:
I didn't want to fork renovate because I would need to always update from upstream and resolve conflicts when `lib/versioning/api.ts` is updated. I thought going with just patching the files would be easy and straight forward, but since they need to be transpiled and added to the Docker image it's not that easy anymore. I am not very familiar with Typescript at all which is another reason why I would like to stay away from any involved solutions.

Any help would be greatly appreciated. Please do tell me if I am not following best practices and there is a better way. I am happy to change my approach.

To summarize:
What is the best approach to build a Docker image from a patched or forked renovate? And make it easy to keep it up to date going forward.

**Comment by viceice** [comment]
Best approach is to submit a pr to include the `rez` versioning to upstrem renovate.

---
Another solution is to create a new project:
- add renovate as dep
- create a `src/versioning/rez.js` file
- create a `src/index.js`
- add `docker-entrypoint.sh` see https://github.com/renovatebot/docker-renovate-full/blob/main/bin/docker-entrypoint.sh

**index.js**
```js
const { getVersionings } = require ('renovate/dist/versioning');
const rez = require ('./versioning/rez');

# register custom versioning, can also be done with managers and datasources
const ds = getVersionings();
ds.set(rez.id, rez.api);

# run renovate
require('renovate/dist/renovate');
```

**Dockerfile**
```Dockerfile
FROM conainerbase/buildpack as base

WORKDIR /usr/src/app

# Build image
#============
FROM base as build

# copy required files
COPY . .

RUN yarn install --frozen-lockfile

# Final image
#============
FROM base as final
# renovate: datasource=github-releases lookupName=containerbase/python-prebuild
RUN install-tool python 3.9.6

# add other tools here


# repare renovate
COPY --from=build /usr/src/app/package.json package.json
COPY --from=build /usr/src/app/src src
COPY --from=build /usr/src/app/node_modules node_modules

COPY docker-entrypoint.sh /usr/local/bin/docker-entrypoint.sh
RUN ln -sf /usr/src/app/src/index.js /usr/local/bin/renovate;
CMD ["renovate"]

USER 1000
```

**Reply by viceice** [reply]
If you like to use typescript, checkout https://github.com/renovatebot/docker-renovate-full for basic's

**Reply by skral** [reply]
@viceice Thank you for the quick reply. I was wondering about creating a PR. Are you open to supporting versioning from some niche package manager? I am happy to invest a couple more hours for clean up if that means this gets merged into upstream. I didn't want to spend time on getting the manager for rez in place though, I have to admit. At least not at the moment.

Will wait for your reply before making a move on the other options.

**Reply by skral** [reply]
The idea of contributing the rez versioning is growing on me. Can you confirm the following steps are ok for me to take, please?

1. Create an official issue for adding rez versioning
1. Fork renovate
1. Create feature branch
1. Write code, tests, and documentation (almost done)
1. Create pull request

**Reply by viceice** [reply]
Sounds good. Basically you should wait for some feedback to the issue before starting implementation. But you've done mostly, so you can also open a pr immediately.
